internal/xds/types/resourceversiontable.go

```diff
@@ -73,3 +73,25 @@ func (t *ResourceVersionTable) AddXdsResource(rType resource.Type, xdsResource t
 
 	t.XdsResources[rType] = append(t.XdsResources[rType], xdsResource)
 }
+
+// AddOrReplaceXdsResource will update an existing resource of rType according to matchFunc or add as a new resource
+// if none satisify the match criteria. It will only update the first match it finds, regardless
+// if multiple resources satisfy the match criteria.
+func (t *ResourceVersionTable) AddOrReplaceXdsResource(rType resource.Type, resource types.Resource, matchFunc func(existing types.Resource, new types.Resource) bool) {
+	if t.XdsResources == nil || t.XdsResources[rType] == nil {
+		t.AddXdsResource(rType, resource)
+		return
+	}
+
+	var found bool
+	for i, r := range t.XdsResources[rType] {
```

Comment:
Oh dang yea I keep forgetting its per loop and not per iteration. I think we're safe because we're doing anything with it that would cause a bug like mutating or adding it somewhere

Reply:
oh nvm `rType` is `types.Resource` so it's copy on value and not returning a pointer. We should be fine here.